[API] More detailed schema information

NobodysNightmare opened the pull request:
## OpenProject Work Packages
- https://community.openproject.org/work_packages/18433
- https://community.openproject.org/work_packages/18570
## Progress
- [x] extend schema of normal properties
- [x] extend schema of linked properties
- [x] include read-only properties
- [x] add specs for all above
- [x] move schema into own endpoint
- [x] merge allowed value related classes
- [x] only show allowed values when embedded in form
  - [x] have specs for that
- [x] actually link to the new schema endpoint

NobodysNightmare committed:
more detailed representation in schema

- represent all properties (that will also stay in the near future)
- show more detailed information
- adapt existing tests (no new ones yet)

NobodysNightmare committed:
accept real title instead of I18N path

NobodysNightmare committed:
increase detail of linked properties in schema

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://feature-api-detailed-schema.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://feature-api-detailed-schema.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
please hound, maintaining readability

NobodysNightmare committed:
let specs consider new schema detail

NobodysNightmare committed:
move schema representers into own directory

NobodysNightmare committed:
extract schema endpoint

- schema representer now represents a schema
- a schema has a limited subset of a WPs functionality
- schema can be created with or without work package

ulferts commented on lib/api/v3/work_packages/form/schema_allowed_priorities_representer.rb:
```diff
@@ -35,6 +35,15 @@ module V3
+                  I18n.t('activerecord.attributes.work_package.priority'),
+                  true,
+                  true,
+                  context)
```
@NobodysNightmare, without looking it up, please tell me what the two `true`s do. This signature is way to large.

NobodysNightmare committed:
restructure schemas with allowed values

- own representer for property schema
- two subclasses for either linked or embedded allowed values
- no special subclasses per represented type

NobodysNightmare committed:
make factory and representer initializer args

they are required, not optional

NobodysNightmare commented on lib/api/decorators/property_schema_representer.rb:
```diff
@@ -32,39 +32,31 @@
-    class AllowedReferenceLinkRepresenter < Roar::Decorator
-      include Roar::JSON::HAL
+    class PropertySchemaRepresenter < ::API::Decorators::Single
+      def initialize(type: nil, name: nil, current_user: nil)
```
@ulferts regarding your comment about the too large interface.
I did some things to improve the situation here:
- have a single base class containing the afforementioned properties (instead two classes + one helper)
- use named parameters to avoid confusion about the meaning of parameters
- kick the bools out of the initializer as they are optional, they can be assigned by the caller after initialization

ulferts replied:
You could get rid of the line below by using [required named parameters](https://github.com/ruby/ruby/blob/ruby_2_1/NEWS#L16-17) (didn't know that myself).

Simply omit the default value

NobodysNightmare replied:
Great! This will reduce the ABC-Size of my methods significantly ^^

NobodysNightmare committed:
clean up refactoring errors

NobodysNightmare committed:
fix styles to please hound

NobodysNightmare committed:
add specs for missing allowed values

NobodysNightmare commented:
It is remarkable how git was able to see so many renames that did not happen, but misses the only real renaming that I performed by moving the `WorkPackageSchemaRepresenter` into a new folder ^^

NobodysNightmare commented:
https://community.openproject.org/work_packages/18570

NobodysNightmare committed:
link the schema resource

NobodysNightmare committed:
style properties as multi line

NobodysNightmare committed:
extract helper for SOME allowed values

only the ones using a simple link for now...

NobodysNightmare committed:
style

NobodysNightmare committed:
OKAY HOUND!

NobodysNightmare committed:
fix wrong member access

NobodysNightmare committed:
style

NobodysNightmare commented on lib/api/v3/work_packages/schema/work_package_schema_representer.rb:
```diff
@@ -0,0 +1,251 @@
+                     writeable: false
+          end
+
+          def self.schema_with_allowed_link(property,
```
I will not address this here, because in a follow up branch this method was moved out and apparently made even worse... I will need to fix it there...

ulferts commented on lib/api/v3/work_packages/schema/work_package_schema_representer.rb:
```diff
@@ -0,0 +1,251 @@
+                          max_length: nil)
+            raise ArgumentError if property.nil? || type.nil?
+
+            title = I18n.t("#{I18N_WORK_PACKAGE}.#{property}") unless title
```
I think you can rely on 

```
WorkPackage.human_attribute_name(property)
```

instead. 

That method also is intelligent enough to use the localisations under `active_record.attributes` as well. Doing that, you can get rid of the duplicate localisations.

ulferts commented on lib/api/v3/work_packages/schema/work_package_schema_representer.rb:
```diff
@@ -0,0 +1,251 @@
+            raise ArgumentError if property.nil? || href_callback.nil?
+
+            type = property.to_s.camelize unless type
+            title = I18n.t("#{I18N_WORK_PACKAGE}.#{property}") unless title
```
You could also place the default values in the method signature. That should reduce the complexity as well.
Please see comment above regarding the localisations.

ulferts commented on lib/api/v3/work_packages/schema/work_package_schema_representer.rb:
```diff
@@ -0,0 +1,251 @@
+
+          schema :percentage_done,
+                 type: 'Integer',
+                 title: I18n.t("#{I18N_WORK_PACKAGE}.done_ratio"),
```
Please see above.

ulferts commented on lib/api/v3/utilities/path_helper.rb:
```diff
@@ -94,6 +94,10 @@ def self.relation(id)
             "#{root}/relations/#{id}"
           end
 
+          def self.schema(project_id, type_id)
```
How will we name the routes to other schemas (e.g. project, version...)

NobodysNightmare replied:
good point... that was not thought-through by me...

NobodysNightmare replied:
`project_schema` and `version_schema` ;-)

ulferts commented on lib/api/v3/work_packages/schema/work_package_schema.rb:
```diff
@@ -33,17 +33,46 @@
 
-            { href: api_v3_paths.version(version.id), title: version.name }
+            @project = project ? project : work_package.project
+            @type = type ? type : work_package.type
```
`@project = project || work_package.project`

is shorter

ulferts commented on lib/api/v3/work_packages/schema/work_package_schema.rb:
```diff
@@ -33,17 +33,46 @@
 
-          self.type = 'Version'
+          def defines_assignable_values?
+            !!@work_package
```
`@work_package.present?` is more speaking.

ulferts commented on lib/api/v3/work_packages/schema/work_package_schema.rb:
```diff
@@ -33,17 +33,46 @@
+          end
+
+          def assignable_versions
+            @work_package.assignable_versions unless @work_package.nil?
```
Why do you define `defines_assignable_values?` and then you do not use it within these methods?

ulferts commented on lib/api/v3/work_packages/schema/work_package_schema_representer.rb:
```diff
@@ -0,0 +1,251 @@
+                          writable: true,
+                          min_length: nil,
+                          max_length: nil)
+            raise ArgumentError if property.nil? || type.nil?
```
Please see above regarding required named parameters.

ulferts commented on lib/api/v3/work_packages/schema/work_package_schema_representer.rb:
```diff
@@ -0,0 +1,251 @@
+                                            href_callback: nil,
+                                            required: true,
+                                            writable: true)
+            raise ArgumentError if property.nil? || href_callback.nil?
```
Please see above regarding required named parameters.

ulferts commented on lib/api/v3/work_packages/schema/work_package_schemas_api.rb:
```diff
@@ -0,0 +1,85 @@
+            # This allows to have a separate schema URL for each kind of different work packages
+            # but with better caching capabilities than simply using the work package id as
+            # identifier for the schema
+            namespace ':id' do
```
@NobodysNightmare like we talked about. Could you check whether it is possible to do something like this:

```
:project_id-:type_id
```

NobodysNightmare replied:
I already wanted to answer that this is not possible, because I could not find anything about this feature here: https://github.com/intridea/grape

However, I decided to at least try your proposal, to be sure that it does not work. Apparently it actually **does** work.

However, I seem to get trouble for URLs like `schemas/foo` now... Will have to investigate here...

NobodysNightmare committed:
Use better localization helper

allows to remove redundant localizations

NobodysNightmare committed:
use required named parameters

NobodysNightmare committed:
doh! place default values into signature

NobodysNightmare committed:
rename path helper

NobodysNightmare committed:
more elegant style

add spec stub for work package schema

ulferts commented:
@NobodysNightmare there are unit tests missing for some of the added classes.

NobodysNightmare committed:
use the full power of grape

ulferts commented:
@NobodysNightmare needs rebase

NobodysNightmare committed:
add specs for WP schema methods passing to WP

NobodysNightmare committed:
more specs for WP schema

NobodysNightmare committed:
WIP

NobodysNightmare committed:
specs and fix for schema resource

NobodysNightmare commented on spec/requests/api/v3/work_packages/work_packages_schemas_resource_spec.rb:
```diff
@@ -0,0 +1,79 @@
+  let(:schema_path) { "/api/v3/work_packages/schemas/#{project.id}-#{type.id}" }
+  let(:current_user) { FactoryGirl.build(:user, member_in_project: project) }
+
+  let(:last_response) { response } # FIXME: why is this needed? Where is this defined elsewhere?
```
Really: What am I doing wrong here?

Without that line the response examples for errors will fail... But I never did this in other integration tests...

NobodysNightmare committed:
style

NobodysNightmare commented:
~~Except my comment about defining `last_response`~~ I'd say I am done...

NobodysNightmare committed:
include Rack

forces/allows the use of last_response instead of response
